### ISSUE TITLE
Environment during post-link.sh / post-link.bat

### ISSUE THREAD

**Issue by rmcgibbo** [conda/conda-build#401]
When conda-build runs the post-link step, the environment in `$PREFIX` doesn't contain an installed python interpreter, even if the package requires python at all steps. I checked this on os-x and windows -- is it intentional?

During post-link, I (unfortunately) need to modify some text files so that the appropriate paths are used to load up DLLs/shared libraries -- and using python itself to modify these files is the most convenient thing. This works fine at the normal user install time, since at that point their conda environment contains python, but it doesn't work at the time post-link is run for testing when running conda-build.

**Comment by stuarteberg** [comment]
I was able to verify this bug with a simple test case and identify the problem.  Note that due to the nature of the bug (explained below), you may or may not be able to reproduce the issue every time.  It depends on the ordering of a particular Python dictionary in the conda internals, which of course is arbitrary from one run to the next.

Here's my test case:

```
$ mkdir aaa
$ cat > aaa/meta.yaml
package:
  name: aaa
  version: 0.1

requirements:
  run:
    - python

$ cat > post-link.sh
echo "Running ${PKG_NAME} post-link in PREFIX=${PREFIX}"
echo "Contents of $PREFIX/bin:"
echo `ls ${PREFIX}/bin`

$ conda build aaa

$ conda create -y -n dummy-env --use-local aaa
```
If the bug strikes, you'll see that `bin` contains no `python` executable.

THE BUG:

The problem is that there is a circular dependency in the python package: `python` depends on `pip`, and `pip` depends on `python`.  When `conda.toposort`[1] is used to determine the order of installation, the circular dependency is detected and the order of installation becomes arbitrary until the dependency cycle is broken.

[1]: https://github.com/conda/conda/blob/master/conda/toposort.py

This could be fixed with a more nuanced algorithm in `toposort()`.

**Comment by tacaswell** [comment]
If the bug is dictionary order related you will probably only be able to reproduce this in >3.4, by default earlier versions do not add salt to the dictionary hash function and the order is random (as compared to insertion order and any 'natural' ordering of the keys) but will always be the same.

**Comment by stuarteberg** [comment]
>by default earlier versions do not add salt to the dictionary hash function and the order is random [...] but will always be the same.

OK, if that's true, then you *should* be able to reproduce the issue using the above test case, every time. :-)

**Comment by tacaswell** [comment]
If you need any convincing, run


```
python -c 'from __future__ import print_function; print([k for k in {1:1, "d": 5, "ab": 1, "b": 4}])'
```

repeatedly under multiple version of python.

**Comment by stuarteberg** [comment]
>If you need any convincing

No, I'm convinced.  I was merely correcting my earlier statement:

>you may or may not be able to reproduce the issue every time

In light of your comment about dictionary order, I was wrong -- it should be easy to reproduce with Python 2.7

**Comment by tacaswell** [comment]
Sorry, there is some internal state here, the 'you' was meant to be general, not 'you' @stuarteberg .

In the past couple of month or two I have come across 4 or 5 bugs (mostly in matplotlib) which are related to this change.  This thread finally made me start thinking about a clean way to demonstrate the change so the example went here.

**Comment by stuarteberg** [comment]
(Regarding the particular circular dependency mentioned above (python/pip), see also: conda/conda#1152, conda/conda#1154.)

**Comment by asmeurer** [comment]
Ah, I didn't realize the pip thing would break all topo sorts. We should definitely fix this.

**Comment by asmeurer** [comment]
FYI you can seed the Python hash by setting the PYTHONHASHSEED environment variable (e.g., for SymPy we always [run the tests in a subprocess](https://github.com/sympy/sympy/blob/d9b59fd03c2a65b9dd783aa0103a98b88121061e/sympy/utilities/runtests.py#L133) so that we can set the seed and display at at the top of the test output).

**Comment by stuarteberg** [comment]
I ran into this issue again today.  I think this issue could be resolved in a reasonably straightforward manner using [`networkx.algorithms.cycles.simple_cycles()`](https://networkx.github.io/documentation/latest/reference/generated/networkx.algorithms.cycles.simple_cycles.html#simple-cycles) within [`conda.toposort()`](https://github.com/conda/conda/blob/master/conda/toposort.py).  But of course, that would add a new dependency to conda (`networkx`).  Is that out of the question?

**Comment by stuarteberg** [comment]
This particular issue can be closed if conda/conda#1614 is merged.
